### PR TITLE
PS-7447 feature: Backport innodb_buffer_pool_in_core_file to 5.7

### PR DESCRIPTION
**Pull request by ldonoso**
<https://jira.percona.com/browse/PS-7447>

PS-7447 : Backport innodb_buffer_pool_in_core_file to 5.7

** Cherry picked 64271f1500e3df80813aaa3ec037d5c4b12a1584 : Bug #27724476	CONTRIBUTION BY FACEBOOK: DETACH INNODB BUFFER POOL BEFORE CREATING CORE

Based on contribution https://bugs.mysql.com/file.php?id=26642&bug_id=90144 from
Facebook, this patch introduces a new dynamic global system variable
innodb_buffer_pool_in_core_file which is ON by default.
When ON, everything should work in a backward compatible way.
Setting it to OFF (for example by passing --skip-innodb-buffer-pool-in-core-file)
means that the user does not want pages allocated for buffer pool to be included in
the core file in the event of dumping core.
If this intention can not be satisfied (for example the OS does not support marking
pages as MADV_DONTDUMP) then an error will be emitted to the log, as soon as this is
determined, and a core_file will be set to OFF, so no core file will be created at all
(which seems to be safer given user's intention).

This patch also:
- introduces linux-version.inc MTR helper file, which makes it easier to
  skip MTR test on old version of Linux.
- refactors all_persisted_variables.test, so that it outputs and uses correct numbers of
  system variables, lowering the effort required to keep it in sync with reality

Reviewed-by: Annamalai Gurusami <annamalai.gurusami@oracle.com>

** Cherry picked f158a8029672cc3ca2b4699b19b3a1e423b081f8 : Bug #27724476	CONTRIBUTION BY FACEBOOK: DETACH INNODB BUFFER POOL BEFORE CREATING CORE

Post-push fix to MTRs which failed when run with --mysqld=--innodb-page-size=4k.
The smaller the page size, the larger the core file even if we exclude buffer pool from it.
This fix sets the threshold for the core file sizes to leave as much margin of error as
possible, which means in equal distance from the core file size for
  --innodb-page-size=4k --skip-innodb-buffer-pool-in-core-file
and
  --innodb-page-size=64k --innodb-buffer-pool-in-core-file
which are respectively the largest core size without buffer pool, and the smallest core
size with buffer pool.

Reviewed-by: Annamalai Gurusami <annamalai.gurusami@oracle.com>

** Cherry picked 99a0de4a51cbe6ebed1e847ea8ceb27e3fbc73bd : BUG#28841164 IF CORE_PATTERN IS COMPLEX: SKIP MYSQLD_CORE_DUMP_WITH...

If core pattern is different from simple 'core' it's not possible for mtr
to find core (and verify size), in such case, just skip tests
innodb.mysqld_core_dump_without_buffer_pool

** Cherry picked fd4554913e2080c19ee30b8e24eecc7699bc140c : Bug#29268761 INNODB.MYSQLD_CORE_DUMP_WITHOUT_BUFFER_POOL TESTS NEED NOT_ASAN.INC TOO

In an ASAN build and with correctly configured core pattern, the tests
mysqld_core_dump_without_buffer_pool,
mysqld_core_dump_without_buffer_pool_dynamic and
mysqld_core_dump_without_buffer_pool_with_resizing
would always fail as AddressSanitizer disables core dumps making it
impossible for tests to locate any generated core files.

Add not_asan.inc in these tests to avoid the problem.

Thanks to Laurynas Biveinis for report and fix.

Change-Id: I36edf0ef2c268c15afaceb7507f8033f36b93c79

** Additional work:

- `/mysql-test/include/excludenoskip.list`: Not applied as this file does not exist in 5.7.

- `/mysql-test/r/all_persisted_variables.result`: Not applied as this file does not exist in 5.7.

- `/mysql-test/suite/innodb/t/mysqld_core_dump_without_buffer_pool_with_resizing.test`: Disable `innodb_disable_resize_buffer_pool_debug`.

- Update `5.7:/mysql-test/suite/perfschema/r/show_sanity.result`

- `/storage/innobase/buf/buf0buf.cc`: Do not use error message
variables as this are not used for warning in the rest of the file.

- `/mysql-test/suite/innodb/r/mysqld_core_dump_without_buffer_pool.test` and `mysqld_core_dump_without_buffer_pool_dynamic.result`: Expected max core size has been decreased to 800 MB as with the previous limit the test passed even when `innodb_buffer_pool_in_core_file` was ON.

### CODE REVIEW

File: storage/innobase/buf/buf0buf.cc [hunk 1]
```diff
@@ -1446,6 +1454,176 @@ buf_block_init(
...
+avoid taking any actions which might trigger a warning or disabling @@core_file.
+*/
+static bool buf_pool_should_madvise = false;
+
+/* Implementation of buf_chunk_t's methods */
+
+/** Advices the OS that this chunk should not be dumped to a core file.
+Emits a warning to the log if could not succeed.
+@return true iff succeeded, false if no OS support or failed */
+bool buf_chunk_t::madvise_dump() {
```
Comment (kamil-holubicki) [inline storage/innobase/buf/buf0buf.cc]: Multiple occurrences: 
Formating. The opening bracket should be in the new line. (what we have here is 8.0-like formating)

File: storage/innobase/buf/buf0buf.cc [hunk 2]
```diff
@@ -1446,6 +1454,176 @@ buf_block_init(
...
+#ifdef HAVE_MADV_DONTDUMP
+  if (madvise(mem, mem_size(), MADV_DODUMP)) {
+      ib::warn() <<
+          "Disabling @@core_file because @@innodb_buffer_pool_in_core_file is disabled, yet madvise(" << mem <<"," << mem_size() << "," <<  "MADV_DODUMP" << ") failed with " << strerror(errno);
```
Comment (kamil-holubicki) [inline storage/innobase/buf/buf0buf.cc]: Multiple occurences:
Line length > 80 characters

File: storage/innobase/include/buf0buf.h [hunk 3]
```diff
@@ -2296,6 +2310,37 @@ struct buf_pool_t{
...
+	/** A wrapper for buf_pool_t::allocator.alocate_large which also advices the
+	  OS that this chunk should not be dumped to a core file if that was requested.
+	  Emits a warning to the log and disables @@global.core_file if advising was
+	  requested but could not be performed, but still return true as the allocation
+	  itself succeeded.
+	  @param[in]	  mem_size  number of bytes to allocate
+	  @param[in/out]  chunk     mem and mem_pfx fields of this chunk will be updated
+	  to contain information about allocated memory region
+	  @return true iff allocated successfully */
+	bool allocate_chunk(ulonglong mem_size, buf_chunk_t *chunk, bool populate);
```
Comment (kamil-holubicki) [inline storage/innobase/include/buf0buf.h]: Formating

File: storage/innobase/include/buf0buf.h [hunk 4]
```diff
@@ -2296,6 +2310,37 @@ struct buf_pool_t{
...
+
+	/** A wrapper for buf_pool_t::allocator.deallocate_large which also advices
+	  the OS that this chunk can be dumped to a core file.
+	  Emits a warning to the log and disables @@global.core_file if advising was
+	  requested but could not be performed.
+	  @param[in]  chunk   mem and mem_pfx fields of this chunk will be used to
+	  locate the memory region to free */
+	void deallocate_chunk(buf_chunk_t *chunk);
```
Comment (kamil-holubicki) [inline storage/innobase/include/buf0buf.h]: Formating

File: storage/innobase/include/sync0sync.h [hunk 5]
```diff
@@ -62,6 +62,7 @@ extern mysql_pfs_key_t	autoinc_mutex_key;
 # ifndef PFS_SKIP_BUFFER_MUTEX_RWLOCK
 extern mysql_pfs_key_t	buffer_block_mutex_key;
 # endif /* PFS_SKIP_BUFFER_MUTEX_RWLOCK */
+extern mysql_pfs_key_t buf_pool_chunks_mutex_key;
```
Comment (kamil-holubicki) [inline storage/innobase/include/sync0sync.h]: formating

File: storage/innobase/sync/sync0sync.cc [hunk 6]
```diff
@@ -51,6 +51,7 @@ mysql_pfs_key_t	autoinc_mutex_key;
 #  ifndef PFS_SKIP_BUFFER_MUTEX_RWLOCK
 mysql_pfs_key_t	buffer_block_mutex_key;
 #  endif /* !PFS_SKIP_BUFFER_MUTEX_RWLOCK */
+mysql_pfs_key_t buf_pool_chunks_mutex_key;
```
Comment (kamil-holubicki) [inline storage/innobase/sync/sync0sync.cc]: formating

File: storage/innobase/buf/buf0buf.cc [hunk 7]
```diff
@@ -1446,6 +1454,176 @@ buf_block_init(
...
+
+/* Implementation of buf_chunk_t's methods */
+
+/** Advices the OS that this chunk should not be dumped to a core file.
+Emits a warning to the log if could not succeed.
+@return true iff succeeded, false if no OS support or failed */
+bool buf_chunk_t::madvise_dump() {
+#ifdef HAVE_MADV_DONTDUMP
+  if (madvise(mem, mem_size(), MADV_DODUMP)) {
+      ib::warn() <<
```
Comment (kamil-holubicki) [inline storage/innobase/buf/buf0buf.cc]: Can we move this (and next) hardcoded message into sql/share/errmsg-utf8.txt like it is done for example for ER_REDO_ENCRYPTION_CANT_BE_CHANGED?

File: storage/innobase/include/ha_prototypes.h [hunk 8]
```diff
@@ -293,6 +293,17 @@ innobase_get_at_most_n_mbchars(
 	ulint data_len,		/*!< in: length of the string in bytes */
 	const char* str);	/*!< in: character string */
 
+/** Checks sys_vars and determines if allocator should mark
+large memory segments with MADV_DONTDUMP
+@return true iff @@global.core_file AND
```
Comment (kamil-holubicki) [inline storage/innobase/include/ha_prototypes.h]: iff -> if

**Review by kamil-holubicki — CHANGES_REQUESTED** [review]

**Comment by kamil-holubicki** [comment]
Please write a proper PR message (https://confluence.percona.com/pages/viewpage.action?pageId=68393728).
Also as the commit message, we should clearly say that this is cherry-picked from commit id: xyz + what was changed/adjusted.

**Review by kamil-holubicki — COMMENTED** [review]
1. Please run Jenkins and paste the link to the job
2. Please clarify code formating with @satya-bodapati. Now it is 8.0-like
3. @satya-bodapati Functionality implementation looks fine for me, but as InnoDB expert please give your opinion

File: storage/innobase/buf/buf0buf.cc [hunk 9]
```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
...
+as on some platforms which do not support madvise() or MADV_DONT_DUMP we need to
+avoid taking any actions which might trigger a warning or disabling @@core_file.
+*/
+static bool buf_pool_should_madvise = false;
+
+/* Implementation of buf_chunk_t's methods */
+
+/** Advices the OS that this chunk should not be dumped to a core file.
+Emits a warning to the log if could not succeed.
+@return true iff succeeded, false if no OS support or failed */
```
Comment (satya-bodapati) [inline storage/innobase/buf/buf0buf.cc]: s/iff/if

File: storage/innobase/buf/buf0buf.cc [hunk 10]
```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
...
+{
+#ifdef HAVE_MADV_DONTDUMP
+	if (madvise(mem, mem_size(), MADV_DODUMP)) {
+		ib::warn() << "Disabling @@core_file because "
+			      "@@innodb_buffer_pool_in_core_file is "
+			      "disabled, yet madvise("
+			   << mem << "," << mem_size() << ","
+			   << "MADV_DODUMP"
+			   << ") failed with " << strerror(errno);
+		return false;
```
Comment (satya-bodapati) [inline storage/innobase/buf/buf0buf.cc]: return(false);

File: storage/innobase/buf/buf0buf.cc [hunk 11]
```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
...
+			   << "MADV_DONTDUMP"
+			   << ") failed with " << strerror(errno);
+		return false;
+	}
+	return true;
+#else  /* HAVE_MADV_DONTDUMP */
+	ib::warn() << "Disabling @@core_file because "
+		      "@@innodb_buffer_pool_in_core_file is disabled, yet "
+		      "MADV_DONTDUMP is not supported on this platform";
+	return false;
```
Comment (satya-bodapati) [inline storage/innobase/buf/buf0buf.cc]: return(false);

File: storage/innobase/buf/buf0buf.cc [hunk 12]
```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
...
+
+/** A wrapper for buf_pool_t::allocator.alocate_large which also advices the OS
+that this chunk should not be dumped to a core file if that was requested.
+Emits a warning to the log and disables @@global.core_file if advising was
+requested but could not be performed, but still return true as the allocation
+itself succeeded.
+@param[in]	mem_size  number of bytes to allocate
+@param[in/out]  chunk     mem and mem_pfx fields of this chunk will be updated
+                          to contain information about allocated memory region
+@return true iff allocated successfully */
```
Comment (satya-bodapati) [inline storage/innobase/buf/buf0buf.cc]: s/iff/if

File: storage/innobase/buf/buf0buf.cc [hunk 13]
```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
...
+bool
+buf_pool_t::allocate_chunk(ulonglong mem_size, buf_chunk_t *chunk,
+			   bool populate)
+{
+	ut_ad(mutex_own(&chunks_mutex));
+	chunk->mem =
+	    allocator.allocate_large(mem_size, &chunk->mem_pfx, populate);
+	if (UNIV_UNLIKELY(chunk->mem == NULL)) {
+		return false;
```
Comment (satya-bodapati) [inline storage/innobase/buf/buf0buf.cc]: return(false);

File: storage/innobase/buf/buf0buf.cc [hunk 14]
```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
...
+	}
+	/* Dump core without large memory buffers */
+	if (buf_pool_should_madvise) {
+		if (!chunk->madvise_dont_dump()) {
+			innobase_disable_core_dump();
+		}
+	}
+	return true;
```
Comment (satya-bodapati) [inline storage/innobase/buf/buf0buf.cc]: return(true);

File: storage/innobase/buf/buf0buf.cc [hunk 15]
```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
...
+Emits a warning to the log if could not succeed.
+@return true iff succeeded, false if no OS support or failed */
+bool
+buf_pool_t::madvise_dump()
+{
+	ut_ad(mutex_own(&chunks_mutex));
+	for (buf_chunk_t *chunk = chunks; chunk < chunks + n_chunks;
+	     chunk++) {
+		if (!chunk->madvise_dump()) {
+			return false;
```
Comment (satya-bodapati) [inline storage/innobase/buf/buf0buf.cc]: return(false);

File: storage/innobase/buf/buf0buf.cc [hunk 16]
```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
...
+			innobase_disable_core_dump();
+		}
+	}
+	allocator.deallocate_large(chunk->mem, &chunk->mem_pfx);
+}
+
+/** Advices the OS that all chunks in this buffer pool instance can be dumped
+to a core file.
+Emits a warning to the log if could not succeed.
+@return true iff succeeded, false if no OS support or failed */
```
Comment (satya-bodapati) [inline storage/innobase/buf/buf0buf.cc]: s/iff/if

File: storage/innobase/buf/buf0buf.cc [hunk 17]
```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
...
+			return false;
+		}
+	}
+	return true;
+}
+
+/** Advices the OS that all chunks in this buffer pool instance should not
+be dumped to a core file.
+Emits a warning to the log if could not succeed.
+@return true iff succeeded, false if no OS support or failed */
```
Comment (satya-bodapati) [inline storage/innobase/buf/buf0buf.cc]: s/iff/if (plz check all places)

File: storage/innobase/buf/buf0buf.cc [hunk 18]
```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
...
+bool
+buf_pool_t::madvise_dont_dump()
+{
+	ut_ad(mutex_own(&chunks_mutex));
+	for (buf_chunk_t *chunk = chunks; chunk < chunks + n_chunks;
+	     chunk++) {
+		if (!chunk->madvise_dont_dump()) {
+			return false;
```
Comment (satya-bodapati) [inline storage/innobase/buf/buf0buf.cc]: same here.. convert all return without braces to return(X); style.. (this is old 5.7 style)

File: storage/innobase/handler/ha_innodb.cc [hunk 19]
```diff
@@ -2062,6 +2066,34 @@ thd_is_select(
 	return(thd_sql_command(thd) == SQLCOM_SELECT);
 }
 
+/** Checks sys_vars and determines if allocator should mark
+  large memory segments with MADV_DONTDUMP
+  @return true iff @@global.core_file AND
+  NOT @@global.innodb_buffer_pool_in_core_file */
+bool innobase_should_madvise_buf_pool() {
```
Comment (satya-bodapati) [inline storage/innobase/handler/ha_innodb.cc]:
```suggestion
/** Checks sys_vars and determines if allocator should mark
large memory segments with MADV_DONTDUMP
@return true if @@global.core_file AND
NOT @@global.innodb_buffer_pool_in_core_file */
bool innobase_should_madvise_buf_pool() {
```

File: storage/innobase/handler/ha_innodb.cc [hunk 20]
```diff
@@ -2062,6 +2066,34 @@ thd_is_select(
...
+	return (test_flags & TEST_CORE_ON_SIGNAL) && !srv_buffer_pool_in_core_file;
+}
+
+/** Make sure that core file will not be generated, as generating a core file
+  might violate our promise to not dump buffer pool data, and/or might dump not
+  the expected memory pages due to failure in using madvise */
```
Comment (satya-bodapati) [inline storage/innobase/handler/ha_innodb.cc]:
```suggestion
/** Make sure that core file will not be generated, as generating a core file
might violate our promise to not dump buffer pool data, and/or might dump not
the expected memory pages due to failure in using madvise */
```

File: storage/innobase/include/buf0buf.h [hunk 21]
```diff
@@ -2090,6 +2096,14 @@ struct buf_pool_t{
 
 	/** @name General fields */
 	/* @{ */
+	BufListMutex chunks_mutex;    /*!< protects (de)allocation of chunks:
+									- changes to chunks, n_chunks are performed
+									while holding this latch,
+									- reading buf_pool_should_madvise requires
+									holding this latch for any buf_pool_t
+									- writing to buf_pool_should_madvise requires
+									holding these latches for all buf_pool_t-s
```
Comment (satya-bodapati) [inline storage/innobase/include/buf0buf.h]: is this properly aligned? cannot makeout from github page

File: storage/innobase/include/buf0buf.h [hunk 22]
```diff
@@ -2296,6 +2310,41 @@ struct buf_pool_t{
...
 
+	/** A wrapper for buf_pool_t::allocator.alocate_large which also advices the
+	  OS that this chunk should not be dumped to a core file if that was requested.
+	  Emits a warning to the log and disables @@global.core_file if advising was
+	  requested but could not be performed, but still return true as the allocation
+	  itself succeeded.
+	  @param[in]	  mem_size  number of bytes to allocate
+	  @param[in/out]  chunk     mem and mem_pfx fields of this chunk will be updated
+	  to contain information about allocated memory region
+	  @return true iff allocated successfully */
```
Comment (satya-bodapati) [inline storage/innobase/include/buf0buf.h]: s/iff/if

Comment (satya-bodapati) [inline storage/innobase/include/buf0buf.h]:
```suggestion
	/** A wrapper for buf_pool_t::allocator.alocate_large which also advices the
	OS that this chunk should not be dumped to a core file if that was requested.
	Emits a warning to the log and disables @@global.core_file if advising was
	requested but could not be performed, but still return true as the allocation
	itself succeeded.
	@param[in]	  mem_size  number of bytes to allocate
	@param[in/out]  chunk     mem and mem_pfx fields of this chunk will be updated
	to contain information about allocated memory region
	@return true iff allocated successfully */
```

Comment (satya-bodapati) [reply storage/innobase/include/buf0buf.h]: Align other comments too this way.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
Looked only for style issues.. Will check functionality next round

File: storage/innobase/include/buf0buf.h (hunk 21 above)
Comment (ldonoso) [reply storage/innobase/include/buf0buf.h]: no, it is not. Thanks for the heads up

**Comment by satya-bodapati** [comment]
@ldonoso Hi! the patch looks good. Looks like the entire 5.7 logic can be applied as is (there are no known changes to buf pool area in 8.0). 

Can you please add jenkins link?

**Comment by ldonoso** [comment]
Please, find the jenkins result at <https://ps57.cd.percona.com/job/percona-server-5.7-param/382/>

File: storage/innobase/include/buf0buf.ic [hunk 23]
```diff
@@ -65,13 +65,13 @@ struct buf_chunk_t{
 	}
 
 	/** Advices the OS that this chunk should not be dumped to a core file.
-	  Emits a warning to the log if could not succeed.
-	  @return true iff succeeded, false if no OS support or failed */
+	Emits a warning to the log if could not succeed.
+	@return true iff succeeded, false if no OS support or failed */
```
Comment (satya-bodapati) [inline storage/innobase/include/buf0buf.ic]: s/iff/if

File: storage/innobase/include/buf0buf.ic [hunk 24]
```diff
@@ -65,13 +65,13 @@ struct buf_chunk_t{
...
 	bool madvise_dump();
 
 	/** Advices the OS that this chunk should be dumped to a core file.
-	  Emits a warning to the log if could not succeed.
-	  @return true iff succeeded, false if no OS support or failed */
+	Emits a warning to the log if could not succeed.
+	@return true iff succeeded, false if no OS support or failed */
```
Comment (satya-bodapati) [inline storage/innobase/include/buf0buf.ic]: s/iff/if

**Review by satya-bodapati — APPROVED** [review]
LGTM! Ok to merge after fixing those last typos

File: storage/innobase/sync/sync0sync.cc (hunk 6 above)
Comment (kamil-holubicki) [reply storage/innobase/sync/sync0sync.cc]: this is not fixed

File: storage/innobase/include/buf0buf.ic (hunk 24 above)
Comment (kamil-holubicki) [reply storage/innobase/include/buf0buf.ic]: this is not fixed

File: storage/innobase/include/buf0buf.ic (hunk 23 above)
Comment (kamil-holubicki) [reply storage/innobase/include/buf0buf.ic]: this is not fixed

**Review by kamil-holubicki — COMMENTED** [review]
Please fix formatting/typo errors.
No other problems detected.

**Review by kamil-holubicki — APPROVED** [review]
LGTM

**Pull request merged**